lib/3scale/backend/storage_async/methods.rb

```diff
@@ -0,0 +1,127 @@
+# frozen_string_literal: true
+
+module ThreeScale
+  module Backend
+    module StorageAsync
+      module Methods
+        # Now we are going to define the methods to run redis commands
+        # following the interface of the redis-rb lib.
+        #
+        # These are the different cases:
+        #   1) Methods that can be called directly. For example SET:
+        #      @redis_async.call('SET', some_key)
+        #   2) Methods that need to be "boolified". These are methods for which
+        #      redis-rb returns a boolean, but redis just returns an integer.
+        #      For example, Redis returns 0 or 1 for the EXISTS command, but
+        #      redis-rb transforms that into a boolean.
+        #   3) There are a few methods that need to be treated differently and
+        #      do not fit in any of the previous categories. For example, SSCAN
+        #      which accepts a hash of options in redis-rb.
+        #
+        # All of this might be simplified a bit in the future using the
+        # "methods" in async-redis
+        # https://github.com/socketry/async-redis/tree/master/lib/async/redis/methods
+        # but there are some commands missing, so for now, that's not an option.
+
+        METHODS_TO_BE_CALLED_DIRECTLY = [
+          :brpoplpush,
+          :del,
+          :exists,
+          :expire,
+          :expireat,
+          :flushdb,
+          :get,
+          :hset,
+          :hmget,
+          :incr,
+          :incrby,
+          :keys,
+          :lindex,
+          :llen,
+          :lpop,
+          :lpush,
+          :lrange,
+          :lrem,
+          :lset,
+          :ltrim,
+          :mget,
+          :ping,
+          :rpush,
+          :sadd,
+          :scard,
+          :setex,
+          :smembers,
+          :srem,
+          :sunion,
+          :ttl,
+          :zcard,
+          :zrangebyscore,
+          :zremrangebyscore,
+          :zrevrange
+        ].freeze
+        private_constant :METHODS_TO_BE_CALLED_DIRECTLY
+
+        METHODS_TO_BE_CALLED_DIRECTLY.each do |method|
+          define_method(method) do |*args|
+            call(method, *args.flatten)
+          end
+        end
+
+        METHODS_TO_BOOLIFY = [
+          :exists?,
+          :sismember,
+          :sadd?,
+          :srem?,
+          :zadd
+        ].freeze
+        private_constant :METHODS_TO_BOOLIFY
+
+        METHODS_TO_BOOLIFY.each do |method|
+          command = method.to_s.delete('?')
```

Comment:
What's the story behind all the methods that got an extra `?` ?

Reply:
Aisonator uses the API from the `redis-rb` gem, which we use in sync mode. In order to work in both sync and async modes, our custom async mode replicates the redis-rb API. The `Methods` class implement this.

The API changed in redis-rb 5, now they addded the methods `:exists?` `:sadd?` and `:srem?` https://github.com/redis/redis-rb/blob/master/CHANGELOG.md#500. We have to adapt our API to reflect that.

For the commands coming with an ending `?` we need to remove it to get the actual command to send to Redis.